Make sure all ports are 16 bit numbers

UdjinM6 opened the pull request:
Ports should always be 16 bit numbers so use proper type (`uint16_t`) and format (`%u`).
Also changing `unsigned short` to `uint16_t` for such cases for consistency.

Reference: https://tools.ietf.org/html/rfc793#section-3.1

UdjinM6 committed:
Make sure all ports are 16 bit numbers

laanwj commented on src/test/netbase_tests.cpp:
```diff
@@ -47,10 +47,10 @@ BOOST_AUTO_TEST_CASE(netbase_properties)
 {
     string hostOut;
-    int portOut = -1;
+    uint16_t portOut = -1;
```
This needs a larger type because it can be -1, which is not equal to 65535.

UdjinM6 replied:
True! Changed to use 0 instead.

laanwj commented:
This probably overlaps a lot with @theuni's net code overhaul.

UdjinM6 committed:
SplitHostPort should fail if provided port is 0 or negative

UdjinM6 commented:
Addressed test issue and changed `SplitHostPort` to return bool and fail if port is 0 or negative. Also added checks to `SplitHostPort` calls to further reassure that ports provided in strings are legit.

theuni commented:
yes, this conflicts somewhat, mostly in net/netbase. I can deal with that though.

More concerning though is the lack over overflow handling on input in a few places. For example:
```
uint16_t port = GetArg("-rpcport", BaseParams().RPCPort());
```
what will "-rpcport=65537" do?

I see a few of those, though admittedly some of them likely already exist. Also looks like a few are fixed, though :)

@UdjinM6 Would you mind first ensuring that all input ports are sanitized (including those coming over the net) before doing the replacement?

UdjinM6 committed:
Validate ports in `GetArg` inputs and add test cases for `TestSplitHost` for ports above upper bound.

UdjinM6 commented:
Thanks for the review @theuni !
If that starts to overlap with your work too much, just tell me - I'm completely fine with closing this PR in favor of your more global work especially if this PR brings some pain in conflicts resolving and such. I would be glad if you any parts of this PR would be useful for your work in some way though.

Meanwhile:
- addressed `GetArg` inputs
- added test cases for `TestSplitHost` for ports above upper bound

I'm not sure which parts of code you were referencing by
> (including those coming over the net)

Could you provide some guidance pls?

theuni commented:
@UdjinM6 by "parts coming over the net", i meant serialized addresses coming from net messages (version, addr, etc). Looks like that's already taken care of.

Concept ACK, and quick review ACK, but It'd be nice to get #8128 in first.

MarcoFalke commented:
Needs rebase after #8128

UdjinM6 committed:
Fix new code after rebase

UdjinM6 commented:
Rebased and (re)added fixes to new code and to code which was moved in #8128.

MarcoFalke commented on src/bitcoin-cli.cpp:
```diff
@@ -148,7 +148,11 @@ static void http_request_done(struct evhttp_request *req, void *ctx)
     std::string host = GetArg("-rpcconnect", DEFAULT_RPCCONNECT);
-    int port = GetArg("-rpcport", BaseParams().RPCPort());
+
+    int portTmp = GetArg("-rpcport", BaseParams().RPCPort());
```
Nit: GetArg returns `int64_t`. I think the goal of the pull was not to cast without check?

UdjinM6 replied:
Agree, will fix

MarcoFalke commented on src/bitcoin-cli.cpp:
```diff
@@ -148,7 +148,11 @@ static void http_request_done(struct evhttp_request *req, void *ctx)
-    int port = GetArg("-rpcport", BaseParams().RPCPort());
+
+    int portTmp = GetArg("-rpcport", BaseParams().RPCPort());
+    if(portTmp <= 0 || portTmp >= 0x10000)
```
Nit: Can you use `!IsValidPort()`, after changing the type of the arg of IsValidPort to int64_t?

UdjinM6 replied:
That would require:
1. include netbase.h
2. move IsValidPort definition to netbase.h
Not sure it worth it tbh.

MarcoFalke commented on src/net.cpp:
```diff
@@ -2069,7 +2075,7 @@ void StartNode(boost::thread_group& threadGroup, CScheduler& scheduler)
         CNetAddr local;
         LookupHost("127.0.0.1", local, false);
-        pnodeLocalHost = new CNode(INVALID_SOCKET, CAddress(CService(local, 0), nLocalServices));
+        pnodeLocalHost = new CNode(INVALID_SOCKET, CAddress(CService(local, (uint16_t)0), nLocalServices));
```
Nit: I think this is too much handholding for developers. I don't see the benefit considering that this will cause unneeded conflicts with other open pulls such as #8542

MarcoFalke commented on src/test/addrman_tests.cpp:
```diff
@@ -92,22 +92,22 @@ BOOST_AUTO_TEST_CASE(addrman_simple)
 
     // Test 2: Does Addrman::Add work as expected.
-    CService addr1 = ResolveService("250.1.1.1", 8333);
+    CService addr1 = ResolveService("250.1.1.1", (uint16_t)8333);
```
Nit: Same here. I don't think this is worth the diff. You already changed the interface of `ResolveService()`, which should be enough, imo.

laanwj commented:
To be honest, I'm not entirely convinced that merging this is worth the risk (smaller integers are easier to overflow, for example).
There's a huge number of changes but no user-facing improvement.

UdjinM6 committed:
Address nits:
- use int64_t to get result from GetArg
- use int64_t for IsPortValid()
- remove manual numbers conversion to uint16_t

MarcoFalke commented:
@laanwj Agree that there are a lot of changes, but I think in the long run we should make sure that arguments are parsed correctly and sanitized. I think this pull improves parsing of port numbers passed in by users.

laanwj commented:
Improving *parsing* is a good thing, if we can factor that out I'd have no problem with it.

laanwj commented:
Closing this for now.
